README.md

```diff
@@ -1,20 +1,506 @@
 # JavaScript Objects
 
-## Objectives
-+ Explain what an object is in JavaScript is
-+ Create an object in JS
-+ Access a value from an object
-+ Add a key-value pair to an object
-+ Delete a key-value pair from an object
-+ Iterate over key-value pairs in an object
-
-
-## What the Student Already Knows
-
-- Variables
-- Data Type- Strings and Integers
-- Functions
-- Booleans, Comparison Operators, Flow Control 
-- Arrays
+## Overview
+JavaScript is an "object-oriented" language. Objects in one form or another are
+an important part of every modern programming language. In this lesson we will
+explore why they are so useful them and how we can use them.
+
+
+## What is an object?
+An object in JavaScript is fundamentally a way to store key-value pairs. Huh?
+What does that mean? Let's look at the following example:
+
+A common problem when writing a chat server is to keep track of connected
+users. Typically each user has an unique name (= "username") and an open
+connection to the server (= "socket"). A server needs to be able to keep track
+of all connected clients. One way using a data structure you already know is
+using an array of username-socket pairs:
+
+```js
+const connections = [['max', conn0], ['alex', conn1], ['james', conn2]];
+```
+
+This totally works: In order to connect a client, you would `push` a new
+`['username', socket]` pair to the `connections` array. In order to disconnect
+a client, you would `splice` the respective pair from the array.
+
+Nevertheless, there are a couple of problems with this approach:
+
+1. Guaranteeing the uniqueness of usernames
+
+    Whenever we add a new client to our `connections` array, we first have to
+    verify that the username isn't already taken by some other client. Thus we
+    would have to iterate over all the previous items in the array. This isn't
+    a huge problem if there are a couple of active clients, but as more and more
+    clients connect to our server, we might run into performance issues.
+
+2. Finding clients
+
+    Similarly finding clients whenever a new client joins or leaves the chat
+    requires us to iterate over completely unrelated items in the `connections`
+    array.
+
+Because of those problems, a couple of really smart people invented "hash maps".
+Hash maps are an interesting data structure that solve a whole lot of issues and
+are useful for modeling a lot of different problems in programming.
+
+All you need to know for now is that JavaScript objects **are** hash maps (to
+make things sound less scary, we are going to use the term "object" from now
+on).
+
+**Advanced:** Actually V8 (Chrome's JavaScript engine) implements a couple of
+different object
+["modes"](http://jayconrod.com/posts/52/a-tour-of-v8-object-representation).
+Depending on the way you use an object, a different implementation might be
+chosen, which can have significant performance implications.
```

Comment:
I'm a _little_ worried that beginners might be intimidated by this section and the above, but I also like that these sections are here. (And flagging them as "advanced" assuages most of my concerns.)